src/rest/resource.cpp

```diff
@@ -518,31 +513,22 @@ void Resource::UpdateDiag(std::string aKey, std::vector<otNetworkDiagTlv> &aDiag
 
 void Resource::Diagnostic(const Request &aRequest, Response &aResponse) const
 {
-    otbrError error = OTBR_ERROR_NONE;
+    otThreadSetReceiveDiagnosticGetCallback(mInstance, &Resource::DiagnosticResponseHandler,
```

Comment:
why setting callback for each `Diagnostic` is necessary?

Reply:
For example, If `factoryreset` is called, the callback may be bind to another function. We should ensure that each time `diagnostic` API in `resource` is called, the response message can be collected by our `resource` handler.

Reply:
After `factoryreset`, `Resource::Init` will run again and sets up the callback ? Is the original code causing real issues?

Reply:
`Resource::Init` will only be run once when we start otbr-agent, so `factoryreset` will set the callback to the default(seems a handler for CLI) . will explain the detail later.

Reply:
> `Resource::Init` will only be run once when we start otbr-agent, then `factoryreset` will set the callback to the default(seems a handler for CLI), I am not sure whether there is another approach that set `otThreadSetReceiveDiagnosticGetCallback` to another function. do you think we do this each time before `diagnostics` API is called is acceptable? or is there an approach that we could detect whether `factoryreset` is called?

I think `factoryreset` re-exec the whole process, so the process restarts as if it was just launched normally. We should not blame `factoryreset` for any issue. If multiple components of ot-br-posix is using `otThreadSetReceiveDiagnosticGetCallback`, maybe they should coordinate with each other. 
Should we use `EventEmitter` so that ot-br-posix calls `otThreadSetReceiveDiagnosticGetCallback` just once and let other components subscribe to the corresponding event using `EventEmitter::On`? @gjc13 @tttttangTH

Reply:
Ok, will try to do like this.